plutus-ledger-api/src/Plutus/V2/Ledger/Tx.hs

```diff
@@ -0,0 +1,145 @@
+
+module Plutus.V2.Ledger.Tx(
+    -- * Transactions
+    TxId (..),
+    ScriptTag (..),
+    RedeemerPtr (..),
+    Redeemers,
+    -- * Transaction outputs
+    TxOut(..),
+    TxOutRef(..),
+    OutputDatum (..),
+    isPubKeyOut,
+    isPayToScriptOut,
+    outAddress,
+    outValue,
+    txOutPubKey,
+    outDatum,
+    outReferenceScript,
+    pubKeyHashTxOut,
+    -- * Transaction inputs
+    TxInType(..),
+    TxIn(..),
+    inRef,
+    inType,
+    inScripts,
+    pubKeyTxIn,
+    scriptTxIn,
+    pubKeyTxIns,
+    scriptTxIns,
+    ) where
+
+import Control.DeepSeq (NFData)
+import Control.Lens
+import Data.Maybe (isJust)
+import GHC.Generics (Generic)
+import Prettyprinter
+
+import PlutusTx qualified
+import PlutusTx.Bool qualified as PlutusTx
+import PlutusTx.Eq qualified as PlutusTx
+
+import Plutus.V1.Ledger.Address
+import Plutus.V1.Ledger.Crypto
+import Plutus.V1.Ledger.Scripts
+import Plutus.V1.Ledger.Tx hiding (TxOut (..), isPayToScriptOut, isPubKeyOut, outAddress, outValue, pubKeyHashTxOut,
+                            txOutDatum, txOutPubKey)
+import Plutus.V1.Ledger.Value
+
+-- | The datum attached to an output: either nothing; a datum hash; or the datum itself (an "inline datum").
+data OutputDatum = NoOutputDatum | OutputDatumHash DatumHash | OutputDatum Datum
+    deriving stock (Show, Eq, Generic)
+    deriving anyclass (NFData)
+
+instance PlutusTx.Eq OutputDatum where
+    {-# INLINABLE (==) #-}
+    NoOutputDatum == NoOutputDatum                = True
+    (OutputDatumHash dh) == (OutputDatumHash dh') = dh PlutusTx.== dh'
+    (OutputDatum d) == (OutputDatum d')           = d PlutusTx.== d'
+    _ == _                                        = False
+
+instance Pretty OutputDatum where
+    pretty NoOutputDatum        = "no datum"
+    pretty (OutputDatumHash dh) = "datum hash: " <+> pretty dh
+    pretty (OutputDatum d)      = "inline datum : " <+> pretty d
+
+-- | A transaction output, consisting of a target address, a value, and optionally a datum hash.
+data TxOut = TxOut {
+    txOutAddress         :: Address,
+    txOutValue           :: Value,
+    txOutDatum           :: OutputDatum,
+    txOutReferenceScript :: Maybe ScriptHash
+    }
+    deriving stock (Show, Eq, Generic)
+    deriving anyclass (NFData)
+
+instance Pretty TxOut where
+    pretty TxOut{txOutAddress, txOutValue, txOutDatum, txOutReferenceScript} =
+                hang 2 $ vsep ["-" <+> pretty txOutValue <+> "addressed to", pretty txOutAddress, "with datum", pretty txOutDatum, "with referenceScript", pretty txOutReferenceScript]
+
+instance PlutusTx.Eq TxOut where
+    {-# INLINABLE (==) #-}
+    (TxOut txOutAddress txOutValue txOutDatum txOutRefScript) == (TxOut txOutAddress' txOutValue' txOutDatum' txOutRefScript') =
+        txOutAddress PlutusTx.== txOutAddress'
+        PlutusTx.&& txOutValue PlutusTx.== txOutValue'
+        PlutusTx.&& txOutDatum PlutusTx.== txOutDatum'
+        PlutusTx.&& txOutRefScript PlutusTx.== txOutRefScript'
```

Comment:
is this more strict than necessary? I thought providing a reference script is kind of optional? i thought the hash+datum is the enough to be the determined factor of an (Eq TxOut). I am not sure about this.

Reply:
Hmm, I don't think so. The `TxOut` is what it is: it's optional to have bits of it, but those still form part of its identity if they're there!